### PR TITLE
Small optimizations

### PR DESCRIPTION
**Pull request by dcbaker**
I decided to start doing some cleanup but got lost in the weeds doing some minor optimizations. There are a couple of small cleanups here like using len() and simplifying the list_to_dict function, but the biggest thing here is rewriting the version checking logic. it's use of `__cmp__` has always annoyed me, and I've always felt that the logic was too complicated, but I've never bothered to look at it closer. When I started running `timeit.timeit` I realized the performance was really bad and decided to fix it. Using `zip()` alone improved the performance a bunch, but getting rid of the `__cmp__` based implementation gives a 3x speedup. I wrote a benchmark from the test case in run_unittests.py, basically just split the version initialization out of the function I timed, over 10,000 iterations the run time dropped from ~15 seconds to ~5 seconds (for 10,000 calls to the function).

[Here's the test in case anyone cares to run it](https://gist.github.com/dcbaker/8d8717515638fa939639613951e9c037)

### CODE REVIEW

File: mesonbuild/mintro.py [hunk 1]
```diff
@@ -343,7 +343,7 @@ def list_projinfo_from_source(sourcedir: str, intr: IntrospectionInterpreter):
     return intr.project_data
 
 def print_results(options, results, indent):
-    if len(results) == 0 and not options.force_dict:
+    if not tresults and not options.force_dict:
```
Comment (ignatenkobrain) [inline mesonbuild/mintro.py]: __[Flake8]__ [F821] undefined name 'tresults' ([link](https://sider.review/gh/repos/19784232/pulls/5312#issue-4002757))

<sub>posted by Sider</sub>

File: mesonbuild/mesonlib.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 import collections
 from enum import Enum
 from functools import lru_cache
+import operator
```
Comment (ignatenkobrain) [inline mesonbuild/mesonlib.py]: __[Flake8]__ [F811] redefinition of unused 'operator' from line 22 ([link](https://sider.review/gh/repos/19784232/pulls/5312#issue-4002756))

<sub>posted by Sider</sub>

File: mesonbuild/mesonlib.py [hunk 3]
```diff
@@ -506,7 +506,6 @@ def detect_vcs(source_dir):
     return None
 
 # a helper class which implements the same version ordering as RPM
-@functools.total_ordering
```
Comment (QuLogic) [inline mesonbuild/mesonlib.py]: This works with `__eq__` and one other comparison operator, not `__cmp__`. Was it really faster (and less code) to re-implement all of the comparison?

Comment (dcbaker) [reply mesonbuild/mesonlib.py]: The benchmark time goes up from ~5 seconds to ~6.3 seconds if if use total_ordering with just `__lt__` and `__eq__`.

The original `__cmp__` implementation is 32 lines, 36 if we added the proper `if isinstance(other, Version) ... else: return NotImplemented` check. this comes out to 45 :shrug:

File: mesonbuild/mesonlib.py [hunk 4]
```diff
@@ -516,7 +514,7 @@ def __init__(self, s):
         # non-alphanumeric separators are discarded
         sequences = [m for m in sequences if not re.match(r'[^a-zA-Z\d]+', m.group(1))]
         # numeric sequences have leading zeroes discarded
-        sequences = [re.sub(r'^0+(\d)', r'\1', m.group(1), 1) for m in sequences]
+        sequences = [int(m.group(1)) if m.group(1).isdigit() else m.group(1) for m in sequences]
```
Comment (jon-turney) [inline mesonbuild/mesonlib.py]: I think this won't handle numeric character sequences corresponding to numbers larger than MAXINT correctly.  Possibly not a practical concern... :)

Comment (dcbaker) [reply mesonbuild/mesonlib.py]: It shouldn't actually be a problem, if we create an int so big that python 3 cant represent it, we'd have run out of memory.

Comment (jon-turney) [reply mesonbuild/mesonlib.py]: Oh yeah, sys.maxint isn't a thing on python3.

**Pull request merged**